### DISCUSSION TITLE
Recording the path of rays as they intersect a geometry

### DISCUSSION THREAD

**Discussion by h-OUS-e** [mitsuba-renderer/mitsuba3#580] [Q&A] [answered]
I am trying to record the paths of rays sampled in multiple directions, but I'm not quite sure how to do that. I thought I could start with the "Script a Renderer" tutorial, but there are a few things I don't understand:

1. If I have 4 surfaces, placed on top of each other, each with a different refractive index, how can I trace the ray origins as it travels from initialization, to the first surface, then to the second surface all the way to the screen, to accumulate a list of origin points from which I can draw a polyline? I currently only know how to intersect initial rays with the scene, and get the rays from that primary intersection after sampling the bsdf of the scene. I don't know how to get the order of each individual sampled ray though.

2. How do I keep this problem differentiable? I know Mitsuba Integrators do that following the Caustic Optimization tutorial, but I wasn't able to rewrite the PTracer in python (which is what I need) as I couldn't inherit the AdjointIntegrator [#559](https://github.com/mitsuba-renderer/mitsuba3/discussions/559). So since I am writing my own loop, I am unsure how to make this differentiable. What I would like to do is to move these surfaces until the origin points of the last intersection are closer to each other (that is to say I would like to optimize the height of these surfaces such that they concentrate incoming light).

Attaching pics of what I am trying to do, and the code block I currently have. Any help is very much appreciated! 

```ruby
# Primary Surface Intersection
si = scene.ray_intersect(rays)
bsdf = si.bsdf(rays)

# Standard BSDF evaluation context for path tracing
bsdf_ctx = mi.BSDFContext()

# Getting Sampler from Camera
loaded_sensor = mi.load_dict(sensor)
sampler = loaded_sensor.sampler().clone()

# Loop iteration counter
i = mi.UInt32(0)

# Initialize the loop state (listing all variables that are modified inside the loop)
loop = mi.Loop(name="", state=lambda: (i, si,))

# while loop(si.is_valid() & (i < ambient_ray_count)):
with dr.suspend_grad():
    # 1. Compute directions of initial surface interaction
    bsdf_sample, bsdf_weight = bsdf.sample(bsdf_ctx, si, sampler.next_1d(),
                                            sampler.next_2d())
    # 2. Spawn a new ray starting at the surface interactions
    rays2= si.spawn_ray(si.to_world(bsdf_sample.wo))
```

![image](https://user-images.githubusercontent.com/66432003/223538679-daa0b27e-db96-4a35-9295-361bced06132.png)
![IMG_6442](https://user-images.githubusercontent.com/66432003/223539085-85e46b70-537d-459f-8cd7-da371d7c1ab0.jpeg)

**Comment by njroussel** [comment, marked as answer]
Hi @h-OUS-e 

1. I'm a bit confused. Every time you have an intersection, you can just save `si.p` to some normal Python list. In addition, of course, you'll want to save a mask to know which rays are still active.

Note: You'll most likely want to use a recorded loop (`mi.Loop`) for performance which is incompatible with what I mentioned above about visualizations. So you can either use a normal Python loop, or not save ray paths during optimization or even have a separate code path just for the visualizations.

2. As long as you keep Mitsuba/DrJit types it will remain a differentiable computation. However I'm uncertain if your specific optimization would require more advances techniques to handle discontinuities (see the beginning of this tutorial). I don't believe it is the case, but I also haven't put much thought into it.

**Reply by h-OUS-e** [reply]
The for loop seems to be working and is generating the desired results, thank you! Attaching an image for reference. 
![image(9)](https://user-images.githubusercontent.com/66432003/224616179-d7c5e38f-13d8-4556-8e1f-26a2fc5dce43.png)

It is slow as you mentioned, and I wanted to see if mi.loop could be used. I don't quite understand though why my kernel crashes with the following code:

```ruby
# Loop iteration counter
i = mi.UInt32(0)

# Initialize the loop state (listing all variables that are modified inside the loop)
loop = mi.Loop(name="", state=lambda: (i, si, sampler))



while loop(si.is_valid() & (i < config['num_of_passes'])):

    # 1. Get BSDF from Surface Interaction and context
    try:
        if len(ray_points) >= 0:
            pass
    except Exception:
        ray_points = [ray_origins_list]

    bsdf = si.bsdf(rays)
    bsdf_ctx = mi.BSDFContext()

    # 2. Sample new ray directions
    
    bsdf_sample, bsdf_weight = bsdf.sample(bsdf_ctx, si, sampler.next_1d(),
                                            sampler.next_2d())
    # 3. Spawn a new ray starting at the surface interactions
    rays2= si.spawn_ray(si.to_world(bsdf_sample.wo))

    # 4. Record points
    ray_points.append(np.array(rays2.o).copy().tolist())

    # 5. Calculate new surface interactions
    si = scene.ray_intersect(rays2)

    # 6. Increase loop Iteration Count
    i+=1
```

The code runs just find without step 4, but introducing step 4 makes it crash for some reason. I am attaching the error below:

```
info 01:31:17.981: Starting interactive window for resource 'c:\Users\ousab\Dropbox (MIT)\ACADEMIA\MIT\Thesis\Code\Experiment_2\230312_rayTracing.py' with controller '.jvsc74a57bd0bb7ec8c706cfe04b661d5a2b4b528d1181cea5c5c9abae69fb221eafb3477399.c:\Users\ousab\miniconda3\envs\mitsuba3\python.exe.c:\Users\ousab\miniconda3\envs\mitsuba3\python.exe.-m#ipykernel_launcher (Interactive)'
info 01:31:18.079: Process Execution: > ~\miniconda3\envs\mitsuba3\python.exe -m pip list
> ~\miniconda3\envs\mitsuba3\python.exe -m pip list
info 01:31:18.435: Starting Jupyter Session startUsingPythonInterpreter, .jvsc74a57bd0bb7ec8c706cfe04b661d5a2b4b528d1181cea5c5c9abae69fb221eafb3477399.c:\Users\ousab\miniconda3\envs\mitsuba3\python.exe.c:\Users\ousab\miniconda3\envs\mitsuba3\python.exe.-m#ipykernel_launcher (Python Path: c:\Users\ousab\miniconda3\envs\mitsuba3, EnvType: Conda, EnvName: 'mitsuba3', Version: 3.10.9 | packaged by conda-forge | (main, Jan 11 2023, 15:15:40) [MSC v.1916 64 bit (AMD64)]) for 'Interactive-1.interactive' (disableUI=false)
info 01:31:18.449: Process Execution: > ~\miniconda3\envs\mitsuba3\python.exe -c "import ipykernel; print(ipykernel.__version__); print("5dc3a68c-e34e-4080-9c3e-2a532b2ccb4d"); print(ipykernel.__file__)"
> ~\miniconda3\envs\mitsuba3\python.exe -c "import ipykernel; print(ipykernel.__version__); print("5dc3a68c-e34e-4080-9c3e-2a532b2ccb4d"); print(ipykernel.__file__)"
info 01:31:18.505: Process Execution: > ~\miniconda3\envs\mitsuba3\python.exe -m ipykernel_launcher --ip=127.0.0.1 --stdin=9003 --control=9001 --hb=9000 --Session.signature_scheme="hmac-sha256" --Session.key=b"8e45782e-6ae7-45a7-83ad-d154897ed37c" --shell=9002 --transport="tcp" --iopub=9004 --f=c:\Users\ousab\AppData\Roaming\jupyter\runtime\kernel-v2-40524xedjqNhJPYBI.json
> ~\miniconda3\envs\mitsuba3\python.exe -m ipykernel_launcher --ip=127.0.0.1 --stdin=9003 --control=9001 --hb=9000 --Session.signature_scheme="hmac-sha256" --Session.key=b"8e45782e-6ae7-45a7-83ad-d154897ed37c" --shell=9002 --transport="tcp" --iopub=9004 --f=c:\Users\ousab\AppData\Roaming\jupyter\runtime\kernel-v2-40524xedjqNhJPYBI.json
info 01:31:18.505: Process Execution: cwd: ~\Dropbox (MIT)\ACADEMIA\MIT\Thesis\Code\Experiment_2
cwd: ~\Dropbox (MIT)\ACADEMIA\MIT\Thesis\Code\Experiment_2
info 01:31:18.996: ipykernel version & path 6.15.0, ~\miniconda3\envs\mitsuba3\lib\site-packages\ipykernel\__init__.py for c:\Users\ousab\miniconda3\envs\mitsuba3\python.exe
info 01:31:20.174: Started Kernel mitsuba3 (Python 3.10.9) (pid: 35496)
info 01:31:20.320: Generated code for 1 = <ipython-input-1-95cdebfc899f> with 501 lines
error 01:31:22.544: Disposing session as kernel process died ExitCode: 3221226505, Reason: c:\Users\ousab\miniconda3\envs\mitsuba3\lib\site-packages\traitlets\traitlets.py:2548: FutureWarning: Supporting extra quotes around strings is deprecated in traitlets 5.0. You can use 'hmac-sha256' instead of '"hmac-sha256"' if you require traitlets >=5.
  warn(
c:\Users\ousab\miniconda3\envs\mitsuba3\lib\site-packages\traitlets\traitlets.py:2499: FutureWarning: Supporting extra quotes around Bytes is deprecated in traitlets 5.0. Use '8e45782e-6ae7-45a7-83ad-d154897ed37c' instead of 'b"8e45782e-6ae7-45a7-83ad-d154897ed37c"'.
  warn(

info 01:31:22.544: Dispose Kernel process 35496.
error 01:31:22.544: Raw kernel process exited code: 3221226505
error 01:31:22.546: Error in waiting for cell to complete [Error: Canceled future for execute_request message before replies were done
	at t.KernelShellFutureHandler.dispose (c:\Users\ousab\.vscode\extensions\ms-toolsai.jupyter-2023.2.1200692131\out\extension.node.js:2:33213)
	at c:\Users\ousab\.vscode\extensions\ms-toolsai.jupyter-2023.2.1200692131\out\extension.node.js:2:52265
	at Map.forEach (<anonymous>)
	at y._clearKernelState (c:\Users\ousab\.vscode\extensions\ms-toolsai.jupyter-2023.2.1200692131\out\extension.node.js:2:52250)
	at y.dispose (c:\Users\ousab\.vscode\extensions\ms-toolsai.jupyter-2023.2.1200692131\out\extension.node.js:2:45732)
	at c:\Users\ousab\.vscode\extensions\ms-toolsai.jupyter-2023.2.1200692131\out\extension.node.js:17:127079
	at ee (c:\Users\ousab\.vscode\extensions\ms-toolsai.jupyter-2023.2.1200692131\out\extension.node.js:2:1552543)
	at dh.dispose (c:\Users\ousab\.vscode\extensions\ms-toolsai.jupyter-2023.2.1200692131\out\extension.node.js:17:127055)
	at hh.dispose (c:\Users\ousab\.vscode\extensions\ms-toolsai.jupyter-2023.2.1200692131\out\extension.node.js:17:134354)
	at process.processTicksAndRejections (node:internal/process/task_queues:96:5)]
warn 01:31:22.546: Cell completed with errors {
  message: 'Canceled future for execute_request message before replies were done'
}
info 01:31:22.546: Cancel all remaining cells true || Error || undefined
```

**Reply by njroussel** [reply]
I think there's something about Windows & Jupyter that is hiding the DrJit error message.

In short, recorded loops are just a recording of the computation they are **not** executed as the Python interpreter reaches them. Appending to the python list cannot be recorded, as it's an operation outside of the DrJit framework: we need to provide the list with an actual value. This requires DrJit to compute the value to be appended at every iteration of the loop.

**Reply by h-OUS-e** [reply]
I see, as a workaround, it seems like I got it to work by defining a class outside the loop, and calling dr.scatter to record ray origins, thank you!

**Comment by h-OUS-e** [comment]
Not sure if I should post this here, but on another note, when the light paths have no where to go, si.p returns [0,0,0] for that particular light path. How can I check if the light path has no where to go, and thus add an arbitrarily large number for that particular light path? Is there a way to do it without going through every single light path sequentially in a for loop?

Trying to make the light paths not connect back to the origin point.
![image](https://user-images.githubusercontent.com/66432003/225517505-7eec2e8d-6629-4edd-a536-62a27a41a857.png)

**Reply by njroussel** [reply]
By "nowhere to go" I assume you're saying that they do not hit a single object in your scene? You can use `si.is_valid()` to get a mask of rays that have not escaped your scene.

**Reply by h-OUS-e** [reply]
Thank you!
